cloudinit/tests/test_upgrade.py

```diff
@@ -0,0 +1,45 @@
+# Copyright (C) 2020 Canonical Ltd.
+#
+# Author: Daniel Watkins <oddbloke@ubuntu.com>
+#
+# This file is part of cloud-init. See LICENSE file for license information.
+
+"""Upgrade testing for cloud-init.
+
+This module tests cloud-init's behaviour across upgrades.  Specifically, it
+specifies a set of invariants that the current codebase expects to be true (as
+tests in ``TestUpgrade``) and then checks that these hold true after unpickling
+``obj.pkl``s from previous versions of cloud-init; those pickles are stored in
+``tests/data/old_pickles/``.
+"""
+
+import operator
+import pathlib
+
+import pytest
+
+from cloudinit.stages import _pkl_load
+from cloudinit.tests.helpers import resourceLocation
+
+
+class TestUpgrade:
+    @pytest.fixture(
+        params=pathlib.Path(resourceLocation("old_pickles")).glob("*.pkl"),
+        scope="class",
+        ids=operator.attrgetter("name"),
+    )
+    def previous_obj_pkl(self, request):
+        """Load each pickle to memory once, then run all tests against it.
+
+        Test implementations _must not_ modify the ``previous_obj_pkl`` which
+        they are passed, as that will affect tests that run after them.
+        """
+        return _pkl_load(str(request.param))
+
+    def test_networking_set_on_distro(self, previous_obj_pkl):
```

Comment:
These tests look great BTW.

Is there any way or test for us to introspect previous attrs on the pkl'd object and compare vs new attrs added in current DataSource implementation as a measure of potential pitfalls on upgrade in referencing attrs/methods that didn't previously exist?

If we had a test that found new attrs on a datasource, then that test could fail if it saw a new attribute listed in the unpickled object and request that a test case is provided for that attr. Not sure if I adequated captured my hope for this suite of tests. I want something to alert my if I've added a potentially upgrade breaking attribute to a datasource and referenced it before it was set during unpickle.

Reply:
> These tests look great BTW.

Thanks!

> Is there any way or test for us to introspect previous attrs on the pkl'd object and compare vs new attrs added in current DataSource implementation as a measure of potential pitfalls on upgrade in referencing attrs/methods that didn't previously exist?

This is tricky for a couple of reasons.

Firstly, because the pickle contains a multitude of objects: there are 10 `BUILD` instructions (which represent instances with state) and 12 `NEWOBJ` instructions (which represent instances which don't use `BUILD`; some of these are instances without state, others will be instances of stdlib types with special handling). So we can't just test the pickled _object_, we have to test all of the instances contained within the pickle.

Secondly, pickles capture instance state: if the instance state is unchanged at instantiation time, it does not follow that it will remain unchanged after a full cloud-init run. Datasources are the obvious example here (and also the root of the pickle tree): instance state which the datasource sets when it fetches metadata, for example, is not set on instantiation. Further, if that not-set-on-instantiation state references an instance itself, then we won't even be testing all of the types of instance that will end up in the pickle.

So, really, we need to perform a full cloud-init run to produce a pickle that is representative of what we can expect to see. If only we had some sort of framework for running cloud-init in instances. :thinking: 

> If we had a test that found new attrs on a datasource, then that test could fail if it saw a new attribute listed in the unpickled object and request that a test case is provided for that attr. Not sure if I adequated captured my hope for this suite of tests. I want something to alert my if I've added a potentially upgrade breaking attribute to a datasource and referenced it before it was set during unpickle.

I had something along these lines in the initial PR: it uses the integration testing framework to launch an instance, waits for boot and pulls obj.pkl. It converts it to a human-readable format (using `pickletools`) and then compares it against a static stored file of the "latest" pickle in the source tree. The idea here is that this test would fail if someone has introduced changes to the pickled state in a PR: the fix would be to update the static pickle representation (which would also act as in-tree documentation of the expected state evolving over time).

This _is_ limited: it only covers the default configuration (on LXD, implicitly), but it would have caught the networking issue. I've pushed up the commits for that to my `pickle2` branch, this is what it looks like compared to this branch: https://github.com/OddBloke/cloud-init/compare/pickle...OddBloke:pickle2 This is nothing like complete; it needs documentation and process around it; but the core idea is there (and works).

Reply:
> Firstly, because the pickle contains a multitude of objects: there are 10 `BUILD` instructions (which represent instances with state) and 12 `NEWOBJ` instructions (which represent instances which don't use `BUILD`; some of these are instances without state, others will be instances of stdlib types with special handling). So we can't just test the pickled _object_, we have to test all of the instances contained within the pickle.

Thank you, that makes sense and lends credence to how painful this is without a full integration test :)

>> ; but the core idea is there (and works).
:+1: